test: Integration tests for handleNewBooking

hariombalhara opened the pull request:
## What does this PR do?

Note: There is some refactor in [the stacked PR](https://github.com/calcom/cal.com/pull/11099). I would be implementing any suggestions there to avoid non trivial conflicts

Adds integration tests for `handleNewBooking`. The tests are added in such a way that it would allow us to refactor the booking flow with lesser worries. I expect not much maintenance to be required for the tests as they mock the constructs that are going to stay there unless we do something drastic. 
<img width="920" alt="Screenshot 2023-09-02 at 3 37 19 PM" src="https://github.com/calcom/cal.com/assets/1780212/0dd1348e-06b0-41ff-bd7b-b7bd886f10ff">

Partially completes #9284. Now the refactor can be done with a bit more confidence.

This was done to work on #7120 with more confidence.

## Detailed Changes
- Abstract out booking tests related code to bookingScenario so that it can be shared b/w getSchedule.test.ts and `handleNewBooking.test.ts`
- Upgraded Vitest to be able to use fixture
- Added vitest-fetch-mock to mock webhook fetch calls.

Following things are mocked
- AppStore CalendarService - We just verify that the respective app's appropriate method has been called.
	- Because it's mocked, Calendar Apps should test their logic themeselves. 
- AppStore VideoApiAdapter - We just verify that the respective app's appropriate method has been called.
	- Because it's mocked, Calendar Apps should test their logic themeselves.  
- `fetch` to verify webhooks
-  Prisma queries


## Run the tests as follows

Watch mode: `yarn tdd handleNewBooking`
One time: `yarn test handleNewBooking`


More improvements in the stacked PR https://github.com/calcom/cal.com/pull/11099

hariombalhara committed:
Add tests

hariombalhara committed:
Fix issues

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into test/handleNewBooking

hariombalhara commented on vitest.config.ts:
```diff
@@ -1,9 +1,11 @@
-      provider: "c8",
+      provider: "v8",
     },
+    testTimeout: 500000,
```
```suggestion
```

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into test/handleNewBooking

hariombalhara committed:
Fixes

hariombalhara committed:
More Fixes

hariombalhara committed:
Remove the changes, just keep the tests

hariombalhara commented on .prettierignore:
```diff
@@ -7,7 +7,6 @@ public
 
 *.lock
 *.log
```
```suggestion
*.log
*.test.ts
```

hariombalhara committed:
Update .prettierignore

hariombalhara commented on apps/web/.prettierignore:
```diff
@@ -1,2 +1 @@
 public/embed
```
```suggestion
public/embed
*.test.ts
```

hariombalhara committed:
Update apps/web/.prettierignore

hariombalhara commented on apps/web/next.config.js:
```diff
@@ -225,6 +225,7 @@ const nextConfig = {
     return config;
   },
   async rewrites() {
+    console.log("process.env.ORGANIZATIONS_ENABLED", process.env.ORGANIZATIONS_ENABLED);
```
```suggestion
```

hariombalhara committed:
Update apps/web/next.config.js

hariombalhara commented on apps/web/test/lib/getSchedule.test.ts:
Utilities from this file are abstracted in `bookingScenarion.ts`

hariombalhara commented on packages/app-store/_utils/getCalendar.ts:
```diff
@@ -29,6 +29,7 @@ export const getCalendar = async (credential: CredentialPayload | null): Promise
   if (calendarType?.endsWith("_other_calendar")) {
     calendarType = calendarType.split("_other_calendar")[0];
   }
+
```
```suggestion
```

hariombalhara committed:
Update packages/app-store/_utils/getCalendar.ts

hariombalhara commented on packages/core/EventManager.ts:
```diff
@@ -120,13 +121,11 @@ export default class EventManager {
       evt["location"] = "integrations:daily";
     }
     const isDedicated = evt.location ? isDedicatedIntegration(evt.location) : null;
```
```suggestion
    const isDedicated = evt.location ? isDedicatedIntegration(evt.location) : null;
    
```

hariombalhara commented on packages/core/EventManager.ts:
```diff
@@ -120,13 +121,11 @@ export default class EventManager {
 
     // If and only if event type is a dedicated meeting, create a dedicated video meeting.
     if (isDedicated) {
       const result = await this.createVideoEvent(evt);
```
```suggestion
      const result = await this.createVideoEvent(evt);
      
```

hariombalhara committed:
Apply suggestions from code review

hariombalhara commented on packages/core/EventManager.ts:
```diff
@@ -120,13 +121,13 @@ export default class EventManager {
     }
     const isDedicated = evt.location ? isDedicatedIntegration(evt.location) : null;
-
+    
```
```suggestion

```

hariombalhara commented on packages/core/EventManager.ts:
```diff
@@ -120,13 +121,13 @@ export default class EventManager {
     if (isDedicated) {
       const result = await this.createVideoEvent(evt);
-
+      
```
```suggestion

```

hariombalhara committed:
Apply suggestions from code review

hariombalhara commented on vitest.config.ts:
```diff
@@ -1,9 +1,11 @@
   test: {
     coverage: {
-      provider: "c8",
+      provider: "v8",
```
Vitest upgrade requires this to be changed.

hariombalhara committed:
Type fixes

hariombalhara committed:
Merge remote-tracking branch 'origin/main' into test/handleNewBooking

hariombalhara commented on apps/web/test/lib/getSchedule.test.ts:
Utilities from this file are abstracted in bookingScenarion.ts. There is no other change in this file.

hariombalhara committed:
Add env variable as needed by unit tests

hariombalhara committed:
Fix typechecks

shivamklr commented:
Testing....

shivamklr committed:
Merge branch 'main' into test/handleNewBooking

shivamklr commented:
Need some time to go through the code and test it. Probably by the End of the week.
@hariombalhara I was wondering if you can add steps to test this in the PR description.

shivamklr committed:
Merge branch 'main' into test/handleNewBooking

shivamklr committed:
Merge branch 'main' into test/handleNewBooking

shivamklr commented on apps/web/test/lib/getSchedule.test.ts:
`yarn test getSchedule` is failing locally. Kindly let me know if I setup something wrong.
![getScheduleTestError](https://github.com/calcom/cal.com/assets/22683064/fdf71d4f-e1e6-4bb5-b266-e5b7598a73f5)

hariombalhara replied:
I think `getSchedule` is known to fail depending on timezones. For now, if it's passing on CI, we should be fine.

shivamklr commented on apps/web/test/utils/bookingScenario.ts:
```diff
@@ -0,0 +1,742 @@
+  });
+}
+
+async function addWebhooks(webhooks: InputWebhook[]) {
```
Do we need `async` here?

```suggestion
function addWebhooks(webhooks: InputWebhook[]) {
```

hariombalhara replied:
Nope, will fix it in the stacked PR.

shivamklr commented on apps/web/test/utils/bookingScenario.ts:
```diff
@@ -0,0 +1,742 @@
+  prismaMock.eventType.findUniqueOrThrow.mockImplementation(eventTypeMock);
+}
+
+async function addBookings(bookings: InputBooking[], eventTypes: InputEventType[]) {
```
```suggestion
function addBookings(bookings: InputBooking[], eventTypes: InputEventType[]) {
```

shivamklr reviewed (dismissed):
Added a few changes, let me know your thoughts on the same.

hariombalhara commented on turbo.json:
```diff
@@ -309,6 +309,7 @@
     "ZOOM_CLIENT_ID",
-    "ZOOM_CLIENT_SECRET"
+    "ZOOM_CLIENT_SECRET",
+    "NEXT_PUBLIC_UNIT_TESTS"
```
Yeah makes sense. I didn't notice that. But also, I am removing it from there now as it doesn't make sense to be there. Unit tests don't use turbo.

hariombalhara commented on vitest.config.ts:
```diff
@@ -1,9 +1,11 @@
 import { defineConfig } from "vitest/config";
 
+process.env.NEXT_PUBLIC_UNIT_TESTS = 1;
```
This is to convey to _baseEmail that you need to collect emails in memory and not send them so that integration tests can verify that. I am thinking of renaming it to NEXT_PUBLIC_INTEGRATION_TESTS

hariombalhara replied:
Reused existing INTEGRATION_TESTS_MODE env variable here.

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.test.ts:
```diff
@@ -0,0 +1,511 @@
+        });
+        mockCalendarToHaveNoBusySlots("googlecalendar");
+
+        createBookingScenario(scenarioData);
```
Nice catch!!
Yeah I guess there was something async in it earlier. I would fix that in this stacked PR https://github.com/calcom/cal.com/pull/11099

hariombalhara commented on apps/web/test/utils/bookingScenario.ts:
```diff
@@ -0,0 +1,742 @@
+    };
+  });
+
+  logger.silly("TestData: Creating EventType", eventTypes);
```
Yeah right now you need to set the log level manually in code like this 
`logger.setSettings({ minLevel: "silly" })`. Don't want to hardcode this, maybe I can pass some argument to vitest and then based on that I can enable this log level, so that it doesn't spam by default.

shivamklr replied:
We can maybe use an .env variable to set the logger level and keep the default to warn and error level.

hariombalhara commented on packages/features/bookings/lib/handleNewBooking.test.ts:
```diff
@@ -0,0 +1,511 @@
+    );
+
+    test(
+      `if booking with Cal Video(Daily Video) fails, booking creation fails with uncaught error`,
```
Yeah absolutely agree.

hariombalhara replied:
Added comment https://github.com/calcom/cal.com/blob/6e9e6dbe1d748bf695af7d0e4e8c215ed705d72e/packages/features/bookings/lib/handleNewBooking.test.ts#L499

hariombalhara committed:
Reuse existing INTEGRATION_TEST_MODE env variable

hariombalhara committed:
Merge remote-tracking branch 'origin/test/handleNewBooking' into test/handleNewBooking

hariombalhara commented:
Thanks @shivamklr for the detailed review 🙏 Addressed all your feedback. All the changes you suggested, I am doing in the stacked PR [here](https://github.com/calcom/cal.com/pull/11099)

joeauyeung committed:
Merge branch 'main' into test/handleNewBooking

zomars commented on packages/features/bookings/lib/handleNewBooking.test.ts:
```diff
@@ -0,0 +1,511 @@
+/**
+ * How to ensure that unmocked prisma queries aren't called?
```
Maybe ensure calling an actual but disposable database? https://github.com/sabinadams/express_sample_app/blob/integration-tests/scripts/run-integration.sh

hariombalhara replied:
For now I have ensured that through `fallbackMockImplementation`

zomars reviewed (approved):
Awesome work @hariombalhara 

Looks ready to merge 

![Ship it](https://media.giphy.com/media/CEIHJK4ympvVe0fAdU/giphy.gif)
